### PR TITLE
Use stored video provider preference to book a visit

### PR DESCRIPTION
**Pull request by jivdhaliwal**
# What
Uses the video provider preference for a Trust (added in https://github.com/madetech/nhs-virtual-visit/pull/443) when booking a new visit

# Why
Removes the reliance on an environment variable to switch between whereby and jitsi. This will make it easier to add new video providers in the future.

# Screenshots

# Notes

### CODE REVIEW

File: pages/api/book-a-visit.js [hunk 1]
```diff
@@ -103,14 +81,22 @@ export default withContainer(
...
+      if (trustErr) {
+        throw trustErr;
       }
 
+      const callIdProvider = new CallIdProvider(
+        trust.videoProvider,
+        body.callTime
+      );
+      const callId = await callIdProvider.generate();
+      let callPassword = ids.generate();
```
Comment (chubberlisk) [inline pages/api/book-a-visit.js:98]: I really like dis - how you've encapsulated this into a class with a switcharoo of providers.

File: src/providers/CallIdProvider.test.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    expect(JitsiCallIdProvider).toHaveBeenCalledWith();
+  });
+
+  it("throws an error if the video provider is not supported", async () => {
+    const prov = new CallIdProvider("notsupported");
+    try {
+      await prov.generate();
+    } catch (e) {
+      expect(e).toEqual("Provider notsupported not supported");
+    }
```
[the remark this answers is not in the document]
Comment (jivdhaliwal) [reply src/providers/CallIdProvider.test.js:27]: I tried `toThrowErrow` but it wasn't throwing the error. I think it's to do with it being async. I wonder if changing it to use a promise chain will help

File: src/providers/CallIdProvider.test.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    const callTime = new Date("2020-05-18 13:00");
+    const prov = new CallIdProvider("whereby", callTime);
+    prov.generate();
+    expect(WherebyCallIdProvider).toHaveBeenCalledWith(callTime);
+  });
+
+  it("returns a call ID for jitsi", async () => {
+    const prov = new CallIdProvider("jitsi");
+    prov.generate();
+    expect(JitsiCallIdProvider).toHaveBeenCalledWith();
```
[the remark this answers is not in the document]
Comment (jivdhaliwal) [reply src/providers/CallIdProvider.test.js:18]: This is checking that we call the Jitsi call id provider with no params. Using `toHaveBeenCalled` would still pass if we passed in any params unexpectedly

File: src/providers/call_id_providers/JitsiCallIdProvider.test.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+jest.mock("../RandomIdProvider", () => {
+  return jest.fn().mockImplementation(() => {
+    return { generate: jest.fn().mockReturnValue("randomId") };
+  });
+});
+
+describe("JitsiCallIdProvider", () => {
+  beforeEach(() => {
+    RandomIdProvider.mockClear();
+  });
```
[the remark this answers is not in the document]
Comment (jivdhaliwal) [reply src/providers/call_id_providers/JitsiCallIdProvider.test.js:13]: Interesting, I've found cases where mocks are not being cleared (the `fetch` mock in book a visit was leaking out to the whereby call id test). I'll investigate further to see what's going on there

File: src/providers/CallIdProvider.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import JitsiCallIdProvider from "./call_id_providers/JitsiCallIdProvider";
```
[the remark this answers is not in the document]
Comment (jivdhaliwal) [reply src/providers/CallIdProvider.js:1]: I went back and forth on this but for consistency lets stick to lower camelcase for paths 👍

File: src/providers/call_id_providers/JitsiCallIdProvider.test.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (jivdhaliwal) [reply src/providers/call_id_providers/JitsiCallIdProvider.test.js]: Removed the mockClears and it worked. Must have been something funky on my side 🤷

**Pull request merged**